### PR TITLE
AQ-159 Implement Serializer public functions

### PR DESCRIPTION
**Pull request by byronlin-bq**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: IonHashDotnet/Serializer.cs [hunk 1]
```diff
@@ -98,7 +142,37 @@ private static void Serializers(IonType type, dynamic value, IIonWriter writer)
...
+                }
+            }
+
+            return bytes;
+        }
+
+        private static byte TQ(IIonValue ionValue)
+        {
+            // https://github.com/amzn/ion-dotnet/issues/13
+            return (byte)((int)ionValue.Type << 4);
```
Comment (simonz-bq) [inline IonHashDotnet/Serializer.cs]: The type is a short, which doesn't need to be cast into an int I think.

Either way, I'd prefer to throw a notImplementedException above the return until the bug is resolved.

File: IonHashDotnet/Serializer.cs [hunk 2]
```diff
@@ -9,52 +9,96 @@
...
         internal void Scalar(IIonValue ionValue)
         {
-            throw new NotImplementedException();
+            this.HandleAnnotationsBegin(ionValue);
+            this.BeginMarker();
+            byte[] scalarBytes = this.GetBytes(ionValue.Type, ionValue.Value, ionValue.IsNull);
+            (byte tq, byte[] representation) tuple = this.ScalarOrNullSplitParts(
+                ionValue.Type,
+                ionValue.IsNull,
+                scalarBytes);
```
Comment (simonz-bq) [inline IonHashDotnet/Serializer.cs]: Does this not fit on one line?

Comment (byronlin-bq) [reply IonHashDotnet/Serializer.cs]: I have line limit set to 120 and it goes over.

File: IonHashDotnet/IonHashDotnet.csproj [hunk 3]
```diff
@@ -11,7 +11,7 @@
     <PackageLicenseFile>LICENSE</PackageLicenseFile>
     <RepositoryUrl>https://github.com/amzn/ion-hash-dotnet</RepositoryUrl>
     <RepositoryType>git</RepositoryType>
-    <OutputType>Library</OutputType>
+    <OutputType>Exe</OutputType>
```
[the remark this answers is not in the document]
Comment (byronlin-bq) [reply IonHashDotnet/IonHashDotnet.csproj]: Reverted

File: IonHashDotnet/Serializer.cs [hunk 4]
```diff
@@ -9,52 +9,96 @@
...
-        private bool hasContainerAnnotation;
+        private bool hasContainerAnnotation = false;
 
         internal Serializer(IIonHasher hashFunction, int depth)
         {
+            this.HashFunction = hashFunction;
+            this.Depth = depth;
         }
 
+        internal int Depth { get; }
```
[the remark this answers is not in the document]
Comment (byronlin-bq) [reply IonHashDotnet/Serializer.cs]: Added private set

File: IonHashDotnet/Serializer.cs (hunk 1 above)
Comment (byronlin-bq) [reply IonHashDotnet/Serializer.cs]: Updated to throw NotImplementedException()

File: IonHashDotnet/IonHashDotnet.csproj (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonz-bq) [reply IonHashDotnet/IonHashDotnet.csproj]: There's still another change though.

**Pull request merged**